thirdparty/autodiff/autodiff/forward/forward.hpp

```diff
@@ -0,0 +1,1807 @@
+
+template<typename T>
+constexpr bool isDual = traits::isDual<plain<T>>::value;
+
+template<typename T>
+constexpr bool isUnaryExpr = traits::isUnaryExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isBinaryExpr = traits::isBinaryExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isTernaryExpr = traits::isTernaryExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isNegExpr = traits::isNegExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isInvExpr = traits::isInvExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isAddExpr = traits::isAddExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isMulExpr = traits::isMulExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isPowExpr = traits::isPowExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isArcTan2Expr = traits::isArcTan2Expr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isHypot2Expr = traits::isHypot2Expr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isHypot3Expr = traits::isHypot3Expr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isNumberDualMulExpr = traits::isNumberDualMulExpr<plain<T>>::value;
+
+template<typename T>
+constexpr bool isNumberDualDualMulExpr = traits::isNumberDualDualMulExpr<plain<T>>::value;
+
+//-----------------------------------------------------------------------------
+// ARE TYPES L AND R EXPRESSION NODES OR NUMBERS, BUT NOT BOTH NUMBERS?
+//-----------------------------------------------------------------------------
+template<typename L, typename R>
+constexpr bool isOperable = (isExpr<L> && isExpr<R>) || (isArithmetic<L> && isExpr<R>) || (isExpr<L> && isArithmetic<R>);
+
+template<typename L, typename C, typename R>
+constexpr bool isOperable3 = (isOperable<L,C> && isOperable<L,R>) || (isOperable<C,L> && isOperable<C,R>) || (isOperable<R,L> && isOperable<R,C>);
+
+namespace traits {
+
+//-----------------------------------------------------------------------------
+// WHAT IS THE VALUE TYPE OF AN EXPRESSION NODE?
+//-----------------------------------------------------------------------------
+
+struct ValueTypeInvalid {};
+
+template<typename T>
+struct ValueType { using type = std::conditional_t<forward::isArithmetic<T>, T, ValueTypeInvalid>; };
+
+template<typename T, typename G>
+struct ValueType<Dual<T, G>> { using type = typename ValueType<plain<T>>::type; };
+
+template<typename Op, typename R>
+struct ValueType<UnaryExpr<Op, R>> { using type = typename ValueType<plain<R>>::type; };
+
+template<typename Op, typename L, typename R>
+struct ValueType<BinaryExpr<Op, L, R>> { using type = common<typename ValueType<plain<L>>::type, typename ValueType<plain<R>>::type>; };
+
+template<typename Op, typename L, typename C, typename R>
+struct ValueType<TernaryExpr<Op, L, C, R>> { using type = common<typename ValueType<plain<L>>::type, common<typename ValueType<plain<C>>::type, typename ValueType<plain<R>>::type>>; };
```

Comment:
In this file are also some pretty long lines.

Reply:
I am surprised that this is not detected by clang-format?

Reply:
I assume that either the autodiff or even the thirdparty folder isn't covered

Reply:
Doesn't run on third part code.

Reply:
This is one reason I believe we should not bundle third party code the way we do.

Reply:
In the first version of my Plugin I used CMake's `ExternalProject` module, to download the dependencies at configure time (but that lead to a problem on one of the test systems). The source files then are in the build-folder. Maybe this could be a cleaner way?

Reply:
If we want to change this (which I still think we should not), it needs to be done in a separate PR. Please leave this as-is for now.